### ISSUE TITLE
[Feature] PS3netserv Support

### ISSUE THREAD

**Issue by horfic** [zurdi15/romm#541]
**Describe the solution you'd like**
Similar with OPL for PS2, PS3netserv support should be added for PS3 games.

That allows games to be run directly from the folder structure with the PS3netserv application
https://github.com/aldostools/webMAN-MOD/wiki/~-PS3-NET-Server

**Comment by gantoine** [comment]
The only way we "support" OPL is just mapping the code names of rom files to IGDB-compatible names. Are you asking if RomM can support the folder structure described in https://github.com/aldostools/webMAN-MOD/wiki/~-PS3-NET-Server#content-directories?

**Comment by horfic** [comment]
hi, i think the mapping would be a good start, but supporting the folderstructure would be best. i guess the same for opl

**Comment by drewdrew21b** [comment]
I think it would be great if the ps3netsrv service could be built into Romm service to simplify how many services must be hosted on top of supporting the underlying folder structure. 

Checking the license for the tooling, they are also using GPL3. 

Additionally this toolset is extremely feature rich, but is now in maintenance mode so it should only require occasional updates going forward. 

If considering this, options could be added under the hood to enable or disable the service via Docker environment variables. 

This is already maintained in docker here:

- https://github.com/shawly/docker-ps3netsrv

**Comment by KompressR** [comment]
I *think* I'm requesting the same thing? Running PS3netserv I am under the impression my file folders need to be in CAPS to support RetroArch's default. I'd prefer to only have one rom library on my server that works for all purposes, so some way to support the ps3netserv file structure would be great. If I'm just unaware of either existing compatibility with the romm file structure working on retroarch through ps3netserv, or if there's a way to set things on ps3netserv to work with the structure, I will gladly do that instead. I don't really care which is compatible with which's structure, as long as some combination works to have just one rom library.

**Comment by drewdrew21b** [comment]
I have my Romm and Ps3netsrv services hosted like this in Docker:

 - Path to rom dir mounted directly. 
   - In this directory create an empty folder for Romm and PS3NetSrv to mount as their base library dirs for each container
 - mount the romm / ps3netsrv folders as their own mount paths within the contsiner
 - From within the docker container shell and within the romm / ps3netsrv folder directories create symlinks to all the folders hosted for each iso type based on the folder structure requirements for romm / ps3netsrv. 

This is annoying because you have to maintain a ton sym links and folders and create new ones each time you onboard new iso types, but it works out well.

**Comment by KompressR** [comment]
Well I can see how that works, but the amount of labour involved isn't something someone automation oriented like myself is willing to put in repeatedly. Ideally there is a simpler implementation. I've run into issues using your method with other containers in the past and while it may not currently be an issue with romm, i'd rather avoid it in the event it does in the future.

I had another question but wasn't sure an issue was something that should be raise about it, but is there a list of platforms that romm can run on? My attempts at others using the PS3 browser have failed although I haven't tried it myself, and I am aware this may be core related as well, however it would be helpful for troubleshooting to know the system you're running on is actually capable, so one doesn't waste their time trying to mess with things that will never help.

**Comment by drewdrew21b** [comment]
Yes it is cumbersome to maintain. I started out with smaller library so it wasn't a big deal in the beginning but it grows increasingly annoying to maintain. 

I have never tried to access my Romm site through the ps3 browser. Not even sure it would work as I chain Romm > Authentik > Plex login credentials for access. I don't know if the ssl certs or even the underlying systems for each of these services would work on ps3 browser. 

I usually use WebMAN Mod to access the remote directories via ps3netsrv directly which is a separate container from my Romm service. I touched this FR because I would like to simplify this configuration and have a one stop shop for Roms on my network as maintaining both services is becoming annoying over time. 

To my first comment, ps3netsrv is basically in maintenance mode. Idk if more critical or difficult features will be added ever or more rom types supported.

**Comment by KompressR** [comment]
I think running two containers with a compatible file structure, or the ability to specify file structure in one or the other would be much simpler than how you're doing things now. Also I'm not set on a ps3 setup yet, so far I'm just running multiman and the latest retroarch ce pkg, but that isn't working out for me, so i've just loaded roms onto the ps3 for now. i have webman installed on the OG ps3, but i haven't been using that lately as I've got my slim in the main room and have been working on friends ps3's mostly lately (still working on hen on one). So again, if I can simply specify a different file structure through using webman, iris, or some other method to share the roms with the ps3, i have no problem commiting to the romm file structure.

**Comment by KompressR** [comment]
> I have my Romm and Ps3netsrv services hosted like this in Docker:
> 
> * Path to rom dir mounted directly.
>   
>   * In this directory create an empty folder for Romm and PS3NetSrv to mount as their base library dirs for each container
> * Mount the romm / ps3netsrv folders as their own mount paths within the container
> * From within the docker container shell and within the romm / ps3netsrv folder directories create symlinks to all the folders hosted for each iso type based on the folder structure requirements for romm / ps3netsrv.
> 
> This is annoying because you have to maintain a ton sym links and folders and create new ones each time you onboard new iso types, but it works out well.

I'm trying to set things up like you've suggested, but I think I'm a little confused. Can you elaborate on how you have your symlinks set up? I created two folders, one for ps3netsrv and one for RomM. the ps3netsrv is setup as would be. the RomM folder only contains a folder named roms, and a symlink from a folder named nes to the ps3netsrv/ROMS/NES folder (one platform to start). RomM doesn't see the files. If i access the RomM docker shell, i can see the symlink, but attempting to browse it goes nowhere. I have verified all permissions. How have you set up your symlinks in the docker shell? will my file structure not work because the actual roms aren't under the folder specified in docker? I can change this of course, I just kept them seperate for troubleshooting. I am hesitant to start trying random symlinks as i've managed to make a mess of that in the past.

**Comment by gantoine** [comment]
Now that we have pkgi support for ps3 systems, and considering the difficulties of implementing ps3netsrv, i'm going to close this issue as will not do. If someone is able to make @drewdrew21b's solution work for them, or wants to propose an alternative, feel free to comment as the issue won't be locked.

**Comment by Alleyway-Jack** [comment]
Eventually I see someone building something either from scratch for ps3 or maybe the dev for something along the lines of Webman (Just an example. Highly doubt it would happen with webman due to a very likely lack of dev interest and memory issues when listing alot of items in the XMB) may add RomM support. Having PKG support is good. But that leaves ISO and folder based game backups out entirely. Ive thrown around the idea of tinkering with trying to get something going for PS3 to integrate with RomM but anything Id manage to try to throw together would be held together by gum and string. If I ever get anything working Ill be sure to share it but dont hold your breath. lol

**Comment by OmniBlade** [comment]
Sad this issue has been closed as will not do, I see there was a lot of talk about making a matching file structure, but a better solution would be for RoMM to implement the ps3netserv protocol since it knows what files are isos, pkgs and such already and could response to the requests directly. Pkgi also is only useful for installing on the internal HD, ps3netserv streams over the network meaning I can use a much larger NAS to store games so it isn't a direct replacement.

**Comment by gantoine** [comment]
> implement the ps3netserv protocol since it knows what files are isos, pkgs and such already and could response to the requests directly

@OmniBlade got any info on the protocol? and any chance you want to try implementing this if you're familiar with it?

**Comment by horfic** [comment]
protocol would probably recreated based on the src https://github.com/aldostools/webMAN-MOD/tree/master/_Projects_/ps3netsrv

not sure if there is a real docu about it, as its the only known service to me which is offering this service

**Comment by OmniBlade** [comment]
@gantoine Not any more than the source code offers as @horfic suggests unfortunately, looks like at its core there are a handful of request types the client can make. My experience is more with C/C++ than python where I've only really made small single file programs to accomplish some specific task and I'm a bit rusty with it. I would probably try and replicate at least the basic undecrypted iso functionality in a command line python program first as a proof of concept before attempting to become familiar with the romm codebase.

There is also https://github.com/xakep666/ps3netsrv-go to study as well.
